File: 2021/20. Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> sStack;
        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                sStack.push(c);
            } else {
                if (sStack.empty() 
                    || (c == ')' && sStack.top() != '(')
                    || (c == '}' && sStack.top() != '{')
                    || (c == ']' && sStack.top() != '[')) {
                    return false;
                }
                sStack.pop();
            }
        }
        return sStack.empty();
    }
};

class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                if (stack.empty()) return false;
                if (c == ')') {
                    if (stack.top() != '(') {
                        return false;
                    }
                } else if (c == '}') {
                    if (stack.top() != '{') {
                        return false;
                    }
                } else if (c == ']') {
                    if (stack.top() != '[') {
                        return false;
                    }
                }
                stack.pop();
            }            
        }
        return stack.empty();
    }
};

File: 2024/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = list()
        mapping = {
            '}': '{',
            ']': '[',
            ')': '('
        }
        for c in s:
            if c in mapping:
                if len(stack) == 0 or stack[-1] != mapping[c]:
                    return False
                else:
                    stack.pop()
            else:
                stack.append(c)
        return len(stack) == 0
